File: TextEditorProject/TextEditor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TextEditorApp
{
    public class TextEditor
    {
        private string content;
        private int cursorPosition;
        private List<string> selections;

        public TextEditor()
        {
            content = "";
            cursorPosition = 0;
            selections = new List<string>();
        }

        public void SetContent(string newContent)
        {
            content = newContent;
            cursorPosition = content.Length;
        }

        public void SetCursorPosition(int position)
        {
            if (position >= 0 && position <= content.Length)
                cursorPosition = position;
            else
                Console.WriteLine("Invalid cursor position.");
        }

        public void AddSelection(string selection)
        {
            selections.Add(selection);
        }

        public void ClearSelections()
        {
            selections.Clear();
        }

        public void DisplayState()
        {
            Console.WriteLine($"Content: {content}");
            Console.WriteLine($"Cursor Position: {cursorPosition}");
            Console.WriteLine($"Selections: {string.Join(", ", selections)}");
        }
    }
}

File: TextEditorProject/src/TextEditor.java
import java.util.ArrayList;
import java.util.List;

public class TextEditor {
    private String content;
    private int cursorPosition;
    private List<String> selections = new ArrayList<>();

    public void setContent(String content) {
        this.content = content;
    }

    public void setCursorPosition(int cursorPosition) {
        this.cursorPosition = cursorPosition;
    }

    public void addSelection(String selection) {
        selections.add(selection);
    }

    public EditorMemento createMemento() {
        return new EditorMemento(content, cursorPosition, new ArrayList<>(selections));
    }

    public void restoreFromMemento(EditorMemento memento) {
        this.content = memento.getContent();
        this.cursorPosition = memento.getCursorPosition();
        this.selections = new ArrayList<>(memento.getSelections());
    }

    public String getContent() {
        return content;
    }

    public int getCursorPosition() {
        return cursorPosition;
    }

    public List<String> getSelections() {
        return selections;
    }
}
